### DISCUSSION TITLE
Can i use with twin

### DISCUSSION THREAD

**Discussion by sankaSanjeeva** [shadcn-ui/ui#558] [Q&A]
Is it possible to use both shadcn and [twin](https://github.com/ben-rogerson/twin.macro) macro in the same react application that was created using [Vite](https://vitejs.dev/)?

**Comment by anatolesns** [comment]
Yes, absolutely — they can coexist in a Vite + React app. They solve different problems and don't conflict architecturally:

- **shadcn/ui** is not a component library you install — it's a CLI that copies component source code (React + Tailwind classes) into your project. Those components just use `cn()` + Tailwind classes under the hood.
- **Twin Macro** lets you write Tailwind classes inside `tw` styled components / props.

They both consume the **same Tailwind config**, so as long as your Tailwind setup serves both, they'll render identical styles.

## The catch: the setup

Twin Macro is the tricky part in Vite, since it's a Babel macro:

1. Install twin.macro, tailwindcss, and twin's Babel plugin: `npm i -D twin.macro babel-plugin-macros tailwindcss`
2. Enable Babel in Vite — usually via `@vitejs/plugin-react` with a Babel config running `babel-plugin-macros`. In `vite.config.ts`:

```ts
react({
  babel: {
    plugins: ['babel-plugin-macros'],
  },
})
```

3. Add `twin` config in `package.json` or a `babel-plugin-macros.config.js`, pointing at your Tailwind config:

```json
{
  "babelMacros": {
    "twin": {
      "config": "./tailwind.config.js",
      "preset": "emotion"
    }
  }
}
```

shadcn's setup on Vite (Tailwind + path aliases) then works on top of that without issues.

## The caveat

Note that shadcn's own setup docs for Vite historically assume Tailwind CSS v4 / PostCSS, while twin.macro is best supported on Tailwind v3. If you're on Tailwind v4, check twin's v4 support status first — that's the most common friction point, not shadcn.

Once both are wired: shadcn components (plain Tailwind classes) and twin-styled components (tw) will share your design tokens, theme colors, and dark mode config — they'll look consistent. 👍
